parcels/particlefile.py

```diff
@@ -31,177 +41,301 @@ def _set_calendar(origin_calendar):
+                    data_dict[var] = np.nan * np.zeros(len(pset))
 
+                i = 0
                 for p in pset:
-                    if _is_particle_started_yet(p, time):
-                        i = p.fileid
-                        self.id[i, self.idx[i]] = p.id
-                        self.time[i, self.idx[i]] = p.time
-                        self.lat[i, self.idx[i]] = p.lat
-                        self.lon[i, self.idx[i]] = p.lon
-                        self.z[i, self.idx[i]] = p.depth
-                        for var in self.user_vars:
-                            getattr(self, var)[i, self.idx[i]] = getattr(p, var)
+                    if p.dt*p.time <= p.dt*time:
+                        for var in self.var_names:
+                            data_dict[var][i] = getattr(p, var)
                         if p.state != ErrorCode.Delete and not np.allclose(p.time, time):
                             logger.warning_once('time argument in pfile.write() is %g, but a particle has time %g.' % (time, p.time))
-
-                for p in first_write:
-                    for var in self.user_vars_once:
-                        getattr(self, var)[p.fileid] = getattr(p, var)
-            else:
-                logger.warning("ParticleSet is empty on writing as array at time %g" % time)
+                        self.maxid_written = np.max([self.maxid_written, p.id])
+                        i += 1
+
+                save_ind = np.isfinite(data_dict["id"])
+                for key in self.var_names:
+                    data_dict[key] = data_dict[key][save_ind]
+
+                if time not in self.time_written:
+                    self.time_written.append(time)
+
+                if len(self.var_names_once) > 0:
+                    first_write = [p for p in pset if (p.id not in self.written_once) and _is_particle_started_yet(p, time)]
+                    data_dict_once['id'] = np.nan * np.zeros(len(first_write))
+                    for var in self.var_names_once:
+                        data_dict_once[var] = np.nan * np.zeros(len(first_write))
+
+                    i = 0
+                    for p in first_write:
+                        self.written_once.append(p.id)
+                        data_dict_once['id'][i] = p.id
+                        for var in self.var_names_once:
+                            data_dict_once[var][i] = getattr(p, var)
+                        i += 1
 
             if not deleted_only:
-                self.idx += 1
                 self.lasttime_written = time
 
-        if sync:
-            self.sync()
+        return data_dict, data_dict_once
+
+    def dump_dict_to_npy(self, data_dict, data_dict_once):
+        """Buffer data to set of temporary numpy files, using np.save"""
+
+        if not os.path.exists(self.tempwritedir):
+            os.mkdir(self.tempwritedir)
+
+        if len(data_dict) > 0:
+            tmpfilename = os.path.join(self.tempwritedir, str(len(self.file_list) + 1) + ".npy")
```

Comment:
Why the +1 here?
In Python iterators start at 0

Reply:
Because it's the next number in line. So if there are already x files in the list, then the next filename is x+1

Reply:
I meant that when list is [], file should be 0.npy,
then you add 1.npy to [0.npy]
no?